server/src/main/java/org/elasticsearch/cluster/service/ClusterApplierService.java

```diff
@@ -216,7 +216,24 @@ public void removeApplier(ClusterStateApplier applier) {
      * Add a listener for updated cluster states
      */
     public void addListener(ClusterStateListener listener) {
-        clusterStateListeners.add(listener);
+        clusterStateListeners.add(new SafeClusterStateListener(listener, threadPool.getThreadContext().newRestorableContext(true)));
```

Comment:
Ah oh no, hang on, this will break `removeListener()`

Reply:
Good catch!

Reply:
Switched to use `removeIf(Predicate)` but that will remove all instances of a listener if it has been added multiple times. Not sure if that's a problem that we expect to see with state change listeners.

Reply:
I expect it isn't, but could we assert that in `addListener`?